### PR TITLE
feat: support for file duplication in Project Panel

### PR DESCRIPTION
**Pull request by LeoBorai**
Release Notes:

- (Added) Support for file duplication

---

## Screenshot

Screenshot 2024-01-30 at 10 47 36 PM

### CODE REVIEW

**Comment by EstebanBorai** [comment]
Hi guys! While reading thought the code I thought I could help by adding a `Duplicate` option that works as a shortcut to copying and pasting the same file in order to duplicate it.

I found out that I cant use `fs` in the context of the action, and I was wondering how FS actions are spawned in order to also update remote while sharing/contributing.

Thanks in advance.

File: crates/project_panel/src/project_panel.rs [hunk 1]
```diff
@@ -872,6 +876,10 @@ impl ProjectPanel {
         }
     }
 
+    fn duplicate(&mut self, _: &Duplicate, cx: &mut ViewContext<Self>) {
+        todo!()
+    }
```
Comment (EstebanBorai) [inline crates/project_panel/src/project_panel.rs:881]: Currently Im here trying to use the `duplicate` method introduced here: https://github.com/zed-industries/zed/pull/7123/files#diff-e13638ad3885e643f14ba8cde1f20fb4955f2cf025ee6b4e7bdb33a0c3f00e50R901-R906

**Review by SomeoneToIgnore — COMMENTED** [review]
I'm not fond of having another action that occupies potentially limited vertical space in the context menus.

Can you give an example of other editors having that action? 
It seems that neither Sublime nor VSCode nor Fleet, no Intellij Platform based IDEs provide it, why should Zed?

Instead, what if we allow the copy+paste combination do the same?

**Comment by SomeoneToIgnore** [comment]
Oh, we do allow to duplicate things with copy+paste invocations, so now I'm totally unsure why this is supposed to be introduced.

**Comment by SomeoneToIgnore** [comment]
Hi.

Thank you for the PR with the help, I had a better look and would want to expand my concerns above:

*  adding `async fn duplicate` onto `fs` level for an action in the part of the UI seems dangerous: `fs` is a relatively low-level crate, trying to lean closer to `std::fs` spirit which has no `duplicate` operation that creates a copy of the file with the `duplicate` extension

* loosing original file extension seems dangerous, and may lead to misaligned file entries in the project panel: it will order multiple duplicates in the same dir as `a.duplicate, b.duplicate, a.js, b.js` which breaks duplicate apart from the originals — `a.js.duplicate` form seems more convenient here

* the existential [question](https://github.com/zed-industries/zed/pull/7123#pullrequestreview-1853108934) from above: duplicate action seems to bring some extra code complexity for something that's not a common thing, why? We can do `cmd-c` `cmd-v` to have something similar already (it adds ` copy` to the end of the file name, before the extension, also keeping the order right)


To answer your question about the FS access, if I've understood it right, you need it in the 
```
fn duplicate(&mut self, _: &Duplicate, cx: &mut ViewContext<Self>) {
    todo!()
}
```
context which has `&self` of type `ProjectPanel`, so all you need to do is call `self.fs` there.


-------------

If you're looking for a better justified project panel change, consider

* https://github.com/zed-industries/zed/issues/5101
* adding an action to toggle join directory-in-directory-in-directory on and off
Which would immensely benefit Java integration:
https://github.com/zed-industries/zed/pull/6935#issuecomment-1913640078

I'm going to close this PR for now, but feel free to reopen if after using `self.fs` and `fs.rs` file unchanged in the new PR, if you insist on having this action.

**Pull request closed without merging**